### DISCUSSION TITLE
MinIO with KES; x509: certificate signed by unknown authority

### DISCUSSION THREAD

**Discussion by skitsanos** [minio/minio#15199] [Q&A] [answered]
Suddenly got on one of the servers `bad certificate` issue. 

```
ERROR Unable to initialize a connection to KES as specified by the shell environment: Post "https://kes-staging:7373/v1/key/create/minio-storage-kms-key": x509: certificate signed by unknown authority
```

Running with env params like this:
```
Environment=MINIO_KMS_KES_ENDPOINT=https://kes-staging:7373/
Environment=MINIO_KMS_KES_KEY_FILE=/home/opc/kes-client/client.key
Environment=MINIO_KMS_KES_CERT_FILE=/home/opc/kes-client/client.crt
Environment=MINIO_KMS_KES_CAPATH=/home/opc/kes-client/server.crt
```

Running `kes key ls -k` on that machine works just fine.

as @aead suggested `the issue is that MinIO does not pick up /home/opc/kes-client/server.crt`

Is there any way to debug it and find the reason why this happening?

## Your Environment

* Version used (`minio --version`): minio version RELEASE.2022-06-25T15-50-16Z (commit-id=bd099f5e71d0ea511846372869bfcb280a5da2f6)
* Server setup and configuration: Oracle Linux
* Operating System and version (`uname -a`): Linux minio-staging 5.4.17-2136.308.9.el7uek.x86_64 #2 SMP Mon Jun 13 20:40:51 PDT 2022 x86_64 x86_64 x86_64 GNU/Linux

**Comment by harshavardhana** [comment]
> Environment=MINIO_KMS_KES_CAPATH=/home/opc/kes-client/server.crt

CAPATH must be `/home/opc/kes-client/` not the file itself.

**Reply by skitsanos** [reply]
making `CAPATH` as folder or file makes no difference. I have one server running with `kes` where it is set to file and it works just fine. What beats me is why identical OS and the same software everywhere produce different results...

**Comment by skitsanos** [comment]
So the actual error message is:

```
Jun 29 19:34:27 minio-staging minio[29699]: ERROR Unable to initialize a connection to KES as specified by the shell environment: Post "https://kes-staging:7373/v1/key/create/minio-storage-kms-key": x509: certificate signed by unknown authority (possibly because of "x509: invalid signature: parent certificate cannot sign this kind of certificate" while trying to verify candidate authority certificate "localhost")
```

**Comment by harshavardhana** [comment, marked as answer]
Okay so this is an invalid cert 

> Jun 29 19:34:27 minio-staging minio[29699]: ERROR Unable to initialize a connection to KES as specified by the shell environment: Post "https://kes-staging:7373/v1/key/create/minio-storage-kms-key": x509: certificate signed by unknown authority (possibly because of "x509: invalid signature: parent certificate cannot sign this kind of certificate" while trying to verify candidate authority certificate "localhost")

Its not related to the CAPATH - its related your server.crt itself which is wrong.

**Comment by skitsanos** [comment]
So indeed it is a problem with the server.crt after all. I just moved kes on the same server where minio is, regenerated certificates and it all works now.
